File: src/types/events.ts
import type { AgentStatus, ToolDefinition, ToolTracePhase } from './trace';

export type { ToolDefinition };

export interface StreamEvent {
  conversationId: string;
  delta: string;
  messageId: string;
}

export interface StreamThinkingEvent {
  conversationId: string;
  delta: string;
  messageId: string;
}

export interface StreamEndEvent {
  conversationId: string;
  messageId: string;
  fullContent: string;
  inputTokens: number;
  outputTokens: number;
}

export interface StreamErrorEvent {
  conversationId: string;
  messageId: string;
  error: string;
}

export interface ToolCallEvent {
  conversationId: string;
  messageId: string;
  toolCallId: string;
  name: string;
  input: Record<string, unknown>;
}

export interface ToolResultEvent {
  conversationId: string;
  messageId: string;
  toolCallId: string;
  result: {
    success: boolean;
    output: string;
    error?: string;
  };
}

export interface ToolTraceEvent {
  conversationId: string;
  sessionId: string;
  turn: number;
  messageId: string;
  toolCallId: string;
  name: string;
  input: Record<string, unknown>;
  phase: ToolTracePhase;
  logicalIndex: number;
  batchId?: number;
  batchIndex?: number;
  isConcurrent?: boolean;
  result?: {
    success: boolean;
    output: string;
    error?: string;
  };
  timestampMs: number;
}

export interface AgentTurnEvent {
  conversationId: string;
  sessionId: string;
  turnCount: number;
  assistantMessageId: string;
}

export interface AgentCompleteEvent {
  conversationId: string;
  sessionId: string;
  messageId: string;
  status: AgentStatus;
  reason: string;
  inputTokens: number;
  outputTokens: number;
}

export interface AgentTurnCompleteEvent {
  conversationId: string;
  sessionId: string;
  turnCount: number;
  status: AgentStatus;
  reason: string;
  inputTokens: number;
  outputTokens: number;
}

File: src-tauri/src/models/events.rs
use serde::{Deserialize, Serialize};

use super::chat::ChatMessage;
use super::tools::{ToolDefinition, ToolResult};

// ─── Tauri Event Payloads ───────────────────────────────────

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamDeltaEvent {
    pub conversation_id: String,
    pub message_id: String,
    pub delta: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamThinkingEvent {
    pub conversation_id: String,
    pub message_id: String,
    pub delta: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamEndEvent {
    pub conversation_id: String,
    pub message_id: String,
    pub full_content: String,
    pub input_tokens: u32,
    pub output_tokens: u32,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct StreamErrorEvent {
    pub conversation_id: String,
    pub message_id: String,
    pub error: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AgentStatus {
    Idle,
    Running,
    Complete,
    Cancelled,
    MaxTurnsReached,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ToolTracePhase {
    Requested,
    Running,
    Completed,
    Failed,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolCallEvent {
    pub conversation_id: String,
    pub message_id: String,
    pub tool_call_id: String,
    pub name: String,
    pub input: serde_json::Value,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolResultEvent {
    pub conversation_id: String,
    pub message_id: String,
    pub tool_call_id: String,
    pub result: ToolResult,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolTraceEvent {
    pub conversation_id: String,
    pub session_id: String,
    pub turn: usize,
    pub message_id: String,
    pub tool_call_id: String,
    pub name: String,
    pub input: serde_json::Value,
    pub phase: ToolTracePhase,
    pub logical_index: usize,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub batch_id: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub batch_index: Option<usize>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub is_concurrent: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub result: Option<ToolResult>,
    pub timestamp_ms: u128,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentTurnEvent {
    pub conversation_id: String,
    pub session_id: String,
    pub turn_count: usize,
    pub assistant_message_id: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentTurnCompleteEvent {
    pub conversation_id: String,
    pub session_id: String,
    pub turn_count: usize,
    pub status: AgentStatus,
    pub reason: String,
    pub input_tokens: u32,
    pub output_tokens: u32,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentCompleteEvent {
    pub conversation_id: String,
    pub session_id: String,
    pub message_id: String,
    pub status: AgentStatus,
    pub reason: String,
    pub input_tokens: u32,
    pub output_tokens: u32,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TracePromptEvent {
    pub conversation_id: String,
    pub session_id: String,
    pub turn: usize,
    pub system_prompt: String,
    pub messages: Vec<ChatMessage>,
    pub tools: Vec<ToolDefinition>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TraceThinkingEvent {
    pub conversation_id: String,
    pub session_id: String,
    pub turn: usize,
}
